docs/development/design/build-images.rst

```diff
@@ -0,0 +1,216 @@
+#. execute git inside the container to clone the repository
+#. analyze and parse files from the repository *outside* the container
+#. create the environment and install docs' dependencies inside the container
+#. execute build commnands inside the container
+#. push the output generated by builds commands to the storage
+
+
+*All* those steps depends on specific commands versions: ``git``, ``python``, ``virtualenv``, ``conda``, etc.
+Currently, we are pinning only a few of them in our Docker images and that have caused issues
+when re-deploying these images with bugfixes: **the images are not reproducible in time**.
+
+.. note::
+
+   The repoducibility of the images will be fixed once
+   https://github.com/readthedocs/readthedocs-docker-images/pull/145 and
+   https://github.com/readthedocs/readthedocs-docker-images/pull/146
+   get merged.
+
+To allow users to pin the image we ended up exposing three images: ``stable``, ``latest`` and ``testing``.
+With that naming, we were able to bugfix issues and add more features
+on each image without asking the users to change the image selected in their config file.
+
+Then, when a completely different image appeared and after testing ``testing`` image enough,
+we discarded ``stable``, old ``latest`` became the new ``stable`` and old ``testing`` became the new ``latest``.
+This produced issues to people pinning their images to any of these names because after this change,
+*we changed all the images for all the users* and many build issues arrised!
+
+
+Goals
+-----
+
+* release a completely new Docker image without forcing users to change their pinned image
+* allow users to stick with an image "forever" (~years)
+* use a ``base`` image with the dependencies that don't change frequently (OS and base requirements)
+* reduce size on builder VM disks by sharing Docker image layers
+* deprecate ``stable``, ``latest`` and ``testing``
+* allow use custom images for particular users/customers by sharing most layers
+* create a small ``nopdf`` image version without LaTeX dependencies for local development
+
+
+New build image structure
+-------------------------
+
+.. Taken from https://github.com/readthedocs/readthedocs-docker-images/blob/master/Dockerfile
+
+* ``ubuntu20-base``
+  * labels
+  * environment variables
+  * system dependencies
+  * install requirements
+  * user requirements
+    * plantuml, imagemagick, rsgv-convert, swig
+    * sphinx-js dependencies
+    * rust
+  * UID and GID
+
+* ``ubuntu20-pdf`` (from ``ubuntu20-base``)
+  * PDF/LaTeX dependencies
+
+* ``ubuntu20`` (from ``ubuntu20-pdf``)
+  * all Python versions (2, 3.6, 3.7, 3.8, 3.9)
+  * conda
+  * future extra user requirements
+  * labels
+
+We will also build a ``nopdf`` version to allow quick testing in local development:
+
+* ``ubuntu20-nopdf`` (from ``ubuntu20-base``)
+  * same as ``ubuntu20`` but based on ``ubuntu20-base`` instead
+
+.. note::
+
+   I don't think it's useful to have ``ubuntu20-py37`` exposed to users,
```

Comment:
Wouldn't that make the build images much smaller? Seems worth doing, no?

Reply:
Each version of Python is ~400mb and we have currently 6 versions, so it will reduce ~2Gb

```bash
# readthedocs/build:latest
docs@71578174d2ac:~/.pyenv$ du -hs versions/*
382M	versions/2.7.18
449M	versions/3.5.10
438M	versions/3.6.12
451M	versions/3.7.9
383M	versions/3.8.6
161M	versions/pypy3.5-7.0.0
docs@71578174d2ac:~/.pyenv$
```

Reducing size on images is good, but we need to think about the complexity that introduces building one image per Python version per OS version and with/without PDF support (24 images with 6 Python versions and 2 OS versions supported, and 28 if we add py3.9). In this scenario, if we need to add something to the base image, we need to rebuild _a lot_ of images. We need to find a balance on this considering the pros and cons for each scenario. What are those pros/cons that you can think about?